pkg/scheduler/plugins/gang/gang_plugin.go

```diff
@@ -0,0 +1,507 @@
+				GetGangNameByPod(pod)))
+		retStatus = framework.NewStatus(framework.Wait)
+		p.ActivateGang(pod, state)
+	case Success:
+		p.AllowGangGroup(pod)
+		retStatus = framework.NewStatus(framework.Success)
+		waitTime = 0
+	}
+	return retStatus, waitTime
+}
+
+// Reserve is the functions invoked by the framework at "reserve" extension point.
+func (p *GangPlugin) Reserve(ctx context.Context, state *framework.CycleState, pod *corev1.Pod, nodeName string) *framework.Status {
+	return nil
+}
+
+// Unreserve
+//(1) handle the timeout gang
+//(2) do nothing when bound failed
+func (p *GangPlugin) Unreserve(ctx context.Context, state *framework.CycleState, pod *corev1.Pod, nodeName string) {
+	gangId := GetNamespaceSplicingName(pod.Namespace,
+		GetGangNameByPod(pod))
+	gang := p.gangCache.GetGangFromCache(gangId)
+	if gang == nil {
+		klog.Errorf("can't find gang, gangName:%v, podName:%v", gangId,
+			GetNamespaceSplicingName(pod.Namespace, pod.Name))
+		return
+	}
+
+	// gang time out
+	if !gang.IsGangResourceSatisfied() {
+		klog.Infof("gang is time out,start to release the assumed resource and add annotations to the "+
+			"gang's children, gangName:%v, podName:%v", gangId, GetNamespaceSplicingName(pod.Namespace, pod.Name))
+		pods, err := p.podLister.List(labels.NewSelector())
+		if err != nil {
+			klog.Errorf("unReserve list  pod err:%v", err.Error())
+			return
+		}
+		for _, podOriginal := range pods {
+			if GetNamespaceSplicingName(pod.Namespace,
+				GetGangNameByPod(pod)) == gangId {
+				pod = podOriginal.DeepCopy()
+				if pod.Annotations == nil {
+					pod.Annotations = map[string]string{}
+				}
+				pod.Annotations[extension.AnnotationGangTimeout] = "true"
+				patchBytes, err := generatePodPatch(podOriginal, pod)
+				if err != nil {
+					klog.Errorf("unReserve generatePodPatch got err:%v", err.Error())
+					continue
+				}
+				if string(patchBytes) == "{}" {
+					continue
+				}
+				err = retry.OnError(
+					retry.DefaultRetry,
+					errors.IsTooManyRequests,
+					func() error {
+						_, err = p.frameworkHandler.ClientSet().CoreV1().Pods(pod.Namespace).Patch(ctx, pod.Name,
+							types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
+						if err != nil {
+							klog.Errorf("unReserve when patch gang timeout annotation to pod:%v, err:%v",
+								GetNamespaceSplicingName(pod.Namespace, pod.Name), err.Error())
+						}
+						return err
+					})
+				if err != nil {
+					klog.Errorf("Failed to patch gang timeout annotation after retry to pod:%v, err:%v",
+						GetNamespaceSplicingName(pod.Namespace, pod.Name), err.Error())
+				}
+			}
+		}
+
+		// release resource of all assumed children of the gang
+		p.frameworkHandler.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
+			if GetNamespaceSplicingName(waitingPod.GetPod().Namespace,
+				GetGangNameByPod(waitingPod.GetPod())) == gangId {
+				klog.Errorf("unReserve rejects the pod name:%v from Gang %s due to timeout",
```

Comment:
unReserve rejects the pod name:%v from Gang %s due to timeout ->. unReserve rejects the pod:%v from Gang:%s due to timeout

Reply:
done